quinn-proto/src/connection/pacing.rs

```diff
@@ -0,0 +1,106 @@
+    time::{Duration, Instant},
+};
+
+/// A simple token-bucket pacer. The bucket starts full and has an adjustable capacity. Once the
+/// bucket is empty, further transmission is blocked. The bucket refills at a rate determined by
+/// the congestion window and estimated round-trip time.
+pub struct Pacer {
+    capacity: u64,
+    tokens: u64,
+    prev: Instant,
+}
+
+impl Pacer {
+    /// Obtains a new [`Pacer`].
+    pub fn new(capacity: u64, now: Instant) -> Self {
+        Self {
+            capacity,
+            tokens: capacity,
+            prev: now,
+        }
+    }
+
+    /// Return how long we need to wait before sending a packet.
+    ///
+    /// If we can send a packet right away, this returns `None`. Otherwise, returns `Some(d)`,
+    /// where `d` is the time before this function should be called again.
+    ///
+    /// The 5/4 ratio used here comes from the suggestion that N = 1.25 in the draft IETF RFC for
+    /// QUIC.
+    pub fn delay(
+        &mut self,
+        smoothed_rtt: Duration,
+        mtu: u16,
+        window: u64,
+        now: Instant,
+    ) -> Option<Instant> {
+        debug_assert_ne!(
+            window, 0,
+            "zero-sized congestion control window is nonsense"
+        );
+        if let Some(time_elapsed) = now.checked_duration_since(self.prev) {
+            self.prev = now;
+            if smoothed_rtt.as_nanos() == 0 {
+                return None;
+            }
+            let elapsed = (time_elapsed.as_nanos() / smoothed_rtt.as_nanos())
+                .try_into()
+                .unwrap_or(u64::max_value());
+
+            let new_tokens = (window.saturating_mul(5) / 4).saturating_mul(elapsed);
+            self.tokens = self.tokens.saturating_add(new_tokens).min(self.capacity);
+        }
+
+        // we disable pacing for extremely large windows
+        if self.tokens >= mtu.into() || window > u32::max_value().into() {
+            return None;
+        }
+
+        let rtt = smoothed_rtt
+            .checked_mul((mtu - self.tokens as u16).into())
+            .unwrap_or_else(|| Duration::new(u64::max_value(), 999_999_999));
+        Some(self.prev + ((rtt / window as u32) / 5) * 4)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn does_not_panic_on_bad_instant() {
+        let old_instant = Instant::now();
+        while Instant::now() <= old_instant {}
+        assert!(Pacer::new(1500, Instant::now())
+            .delay(Duration::from_micros(0), 0, 1, old_instant)
+            .is_none());
+        assert!(Pacer::new(1500, Instant::now())
+            .delay(Duration::from_micros(0), 1600, 1, old_instant)
+            .is_some());
+        assert!(Pacer::new(1500, Instant::now())
+            .delay(Duration::from_micros(0), 1500, 3000, old_instant)
+            .is_none());
+    }
+    #[test]
+    fn computes_pause_correctly() {
+        let old_instant = Instant::now();
+        while Instant::now() <= old_instant {}
```

Comment:
Rather than trying to massage the value of `Instant::now()`, just add a fixed `Duration` to it manually. That's simpler and provides better reproducibility.